#과제 5

###1. 클래스와 모듈의 공통점과 차이점에 대해 설명하시오.

공통점으로는 모듈과 클래스는 모두 프로그램의 재사용을 위한 코드이다. 기능별로 독립하여 검사와 안정성에 유리하며,
각 클래스나 모듈별로 다른 이름공간을 가지므로 독립된 프로그래밍이 가능하다

모듈은 파일 단위로 이름공간을 구성하지만, 클래스는 클래스 영역과 인스턴스 영역에 이름공간을 구성한다.

In [2]:
# 2. 다형성에 대해 설명하고 다형성을 보여주는 자신만의 파이썬 코드 예제를 제시하시오.

#다형성이란 상속 관계 내의 다른 클래스들의 인스턴스들이 같은 멤버 함수 호출에 대해 각각 다르게 반응하도록 하는 기능

class Department:
    def slogan(self):
        print 'good!'
        
class Computer(Department):
    def slogan(self):
        print 'fighting Computer!'
        
class Mechatronics(Department):
    def slogan(self):
        print 'fighting Meca'
        
class Architecture(Department):
    def slogan(self):
        print 'fighting Architecture'
        
class Design(Department):
    def slogan(self):
        print 'fighting Design'

for x in (Computer(), Mechatronics(), Architecture(),Design()):
    x.slogan()



fighting Computer!
fighting Meca
fighting Architecture
fighting Design


다형성이란 상속 관계 내의 다른 클래스들의 인스턴스들이 같은 멤버 함수 호출에 대해 각각 다르게 반응하도록 하는 기능이며 코드를 보면 slogan()이라는 멤버 함수를 통해 모든 클래스들의 각자 다르게 구현된 slogan()을 호출 할수 있습니다. 즉 같은 이름의 함수더라도 그 기능은 다르게 구현할수 있습니다.

In [19]:
'''3.다음 각 요구사항 모두를 만족시키는 Counter 클래스를 코딩하시오 (정답을 각 요구사항별로 입력할 필요 없이 3번 문제에 대해 1개의 클래스 
정의 코드를 제시하면 된다.)

•요구사항 1. 생성자에 count 값과 step 값을 인자로 받을 수 있다.
◾count: Counter 인스턴스가 지니는 초기 정수 값
◾step: Counter 인스턴스의 count가 증가되는 증분 (defalt 값: 1)

'''
class Counter:
    def __init__(self,count,step=1):
        self.count=count
        self.step=step
    def incr(self):
         self.count+=self.step
    def __str__(self):
        return '[Count(step:%s)]%s'%(self.step,self.count)
    def __call__(self):
        return self.incr()
    def __add__(self,other):
        return Counter(self.count+other)
    def __sub__(self,other):
        return Counter(self.count-other,2)
    def __cmp__(self,other):
        return self.count-other
c=Counter(10)
d=Counter(10,2)
print c
print d
c.incr()
d.incr()
print c
print d
c()
d()
print c
print d
c=c+5
d=d-5
print c
print d
print c>10
print d>10
print c<10
print d<10
print c==10
print d!=9



[Count(step:1)]10
[Count(step:2)]10
[Count(step:1)]11
[Count(step:2)]12
[Count(step:1)]12
[Count(step:2)]14
[Count(step:1)]17
[Count(step:2)]9
True
False
False
True
False
False


생성자 함수 __init__가 인자를 받을 수 있도록 하였고, step의 default 값을 1로 놓으라고 하였으므로 step=1을 하였습니다.  print를 위한 __str__ 함수와 count 값 증가를 위한 incr 함수는 count값과 step값을 더했고, 호출 가능형으로 만들기 위한 call 함수 및 연산을 위한 add, sub(2번째 step임을 나타내기 위해 step값에 2를 넣어주었습니다.) 함수와 비교를 위한 cmp 함수를 정의하였습니다.

In [20]:
'''4.다음은 내장 자료형 list를 서브클래싱하여 만든 MySet 클래스 정의 내용이다. 다음 클래스 정의에서
__init__(), __str()__(), elimicate_duplicate()의 세 개의 메소드 코드 내용을 자신이 다른 사람에게 가르친다고 생각하며 설명해보시오.
•MySet은 집합(Set) 자료형을 정의하려는 의도하에 만들어진 클래스이다.
'''
class MySet(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)

    
if __name__ == "__main__":
    s = MySet([1, 2, 2, 3])
    print s
    t = MySet([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])
    print t


MySet: {1 ,2 ,3}
MySet: {2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}


먼저 init()함수는 생성자함수로서 객체에 인자로 받은 리스트의 모든 원소를 for문을 통해 하나하나 추가합니다. 그리고 최종적으로 eliminate_duplicate() 함수가 리스트를 호출 하도록 합니다. 여기서 eliminate_duplicate() 함수는 새로운 s리스트를 만들어 원소들을 하나씩 저장하지만 검사를 하여 중복되지 않는 것만 저장합니다. 이후 str()은 어떤식으로 print 될것인가를 정하는데 먼저 result에  MySet: {를 저장하고 그뒤에 받은 리스트의 원소들을 하나하나 추가 해줍니다. 그리고 마지막 띄어쓰기와 ,를 없애주고 }를 추가해준 result값을 리턴하여줍니다.

In [44]:
'''5. 4번 문제에 정의된 MySet 클래스에 메소드를 추가하여 다음 각 요구사항 모두를 만족시키는 코딩을 제시하시오
•정답을 각 요구사항별로 입력할 필요 없이 요구사항 3개 전체에 대해 1개의 MySet 클래스 정의 코드를 제시하고 주석등으로 해당 코드를 설명한다.

•요구사항 1. | 연산으로 두 집합의 합집합을 반환한다.
>>> u = s | t 
 >>> print u 
 MySet: {1, 2, 3, 4, 5, 6, 7, 8, 9} 

•요구사항 2. & 연산으로 두 집합의 교집합을 반환한다.
>>> u = s & t 
 >>> print u 
 MySet: {2, 3} 

•요구사항 3. - 연산으로 두 집합의 차집합을 반환한다.
>>> s = MySet([1, 2, 3]) 
 >>> t = MySet([3, 4, 5]) 
 >>> u = s - t 
 >>> print u 
 MySet: {1} 
'''
class MySet(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)
            
            
    #-------------------
    def union(self, A):  #같은 값이 없을 때만 리스트에 값을 추가해줍니다.
        for x in A:
            if x not in self:
                self.append(x)
        return MySet(self)
    def intersection(self, B): # 자신의 리스트와 다른 인스턴스의 리스트의 값 둘다 있는 원소만 새로운 w리스트에 추가해줍니다
        w=[]
        for x in self:
            if x in B:
                w.append(x)
        return MySet(w)
    def difference(self, C): #자신의 리스트중 다른 인스턴스의 리스트와 겹치지 않은 원소만 w에 추가해 줍니다.
        w=[]
        for x in self:
            if x not in C:  
                w.append(x)
        return MySet(w)
    def __or__(self,D):    # '|'연산자를 사용하기 위해 해사용 ,인자 값으로 자신과 다른 인스턴스를 받아 nion()함수값으로 반환해줍니다.
        return self.union(D)
    def __sub__(self,E):  #‘-’에 대한 연산자 오버로딩을 정의 합니다. 인자를 받아 difference()함수값으로 출력합니다
        return self.difference(E)
    def __and__(self,F):  #‘&’에 대한 연산자 오버로딩을 정의 합니다. 인자를 받아 intersection()함수값을 반환 해줍니다.
        return self.intersection(F)
    
if __name__ == "__main__":
    s = MySet([1, 2, 2, 3])
    print s
    t = MySet([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])
    print t
    

u=s|t
print u

s = MySet([1, 2, 2, 3])
t = MySet([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])

u=s & t
print u

s = MySet([1, 2, 3]) 
t = MySet([3, 4, 5]) 
u = s - t 
print u 

MySet: {1 ,2 ,3}
MySet: {2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}
MySet: {1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}
MySet: {2 ,3}
MySet: {1 ,2}


마지막에 s = MySet([1, 2, 3]) ,t = MySet([3, 4, 5])을 넣을 경우 {1} 이나온다고 되있는데 이는 잘못쓰신 것 같습니다. 차집합은 1,2 가 나와야 하는 것 같습니다

In [45]:
'''6. 5번 문제에서 정의한 MySet 클래스에 대해 다음 예제를 수행하면 오류없이 올바르게 동작하는 것을 확인할 수 있다.
다음 예제 내에 있는 len(), bool() 내장함수와 in 키워드 사용 예제가 별다른 메소드 정의를 하지 않았는 데도 올바르게 수행되는 이유를 설명하시오.
> \>\>\> s = MySet([1, 2, 3, 4, 5, 6])  <br/>
> \>\>\> print len(s)  <br/>
> 6 <br/>
> \>\>\> print bool(s)  <br/>
> True <br/>
> \>\>\> print 2 in s  <br/>
> True <br/>
'''

class MySet(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)
            
            
    #-------------------
    def union(self, A):  #같은 값이 없을 때만 리스트에 값을 추가해줍니다.
        for x in A:
            if x not in self:
                self.append(x)
        return MySet(self)
    def intersection(self, B): # 자신의 리스트와 다른 인스턴스의 리스트의 값 둘다 있는 원소만 새로운 w리스트에 추가해줍니다
        w=[]
        for x in self:
            if x in B:
                w.append(x)
        return MySet(w)
    def difference(self, C): #자신의 리스트중 다른 인스턴스의 리스트와 겹치지 않은 원소만 w에 추가해 줍니다.
        w=[]
        for x in self:
            if x not in C:  
                w.append(x)
        return MySet(w)
    def __or__(self,D):    # '|'연산자를 사용하기 위해 해사용 ,인자 값으로 자신과 다른 인스턴스를 받아 nion()함수값으로 반환해줍니다.
        return self.union(D)
    def __sub__(self,E):  #‘-’에 대한 연산자 오버로딩을 정의 합니다. 인자를 받아 difference()함수값으로 출력합니다
        return self.difference(E)
    def __and__(self,F):  #‘&’에 대한 연산자 오버로딩을 정의 합니다. 인자를 받아 intersection()함수값을 반환 해줍니다.
        return self.intersection(F)
    
s = MySet([1, 2, 3, 4, 5, 6])  
print len(s)
print bool(s)
print 2 in s

6
True
True


len()과 bool() 내장함수에 이미 저장 되어 있고 in 키워드는 리스트 자료형에 정의된 메소드이기 때문입니다. 
MySet 클래스는 리스트 클래스를 상속받아 정의한 클래스이므로 부모클래스인 리스트 클래스의 메소드들을 사용할 수 있습니다.

In [71]:
'''7. 215 = 32768 의 각 자리수를 더하면 3 + 2 + 7 + 6 + 8 = 26 입니다.

21000의 각 자리수를 모두 더하면 얼마입니까?
'''
import math


t= math.pow(2, 1000)
t=str(int(t))
i=0
for x in t:
    i= i+int(x)
    
print i

1366


2의 1000승을 pow()함수로 구한후 이값은 실수기 때문에 뒤에 .0 이 붙어 있습니다. 이를 제거하기위해 int형으로 만든 후 이를 하나씩 더하기 쉽게 하기위해 다시 문자열로 바꾸었습니다.이것을 for문을 통해 하나씩 int형으로 바꾸어 주면서 더했습니다.

In [77]:
'''8. 1부터 5까지의 숫자를 영어로 쓰면 one, two, three, four, five 이고,
각 단어의 길이를 더하면 3 + 3 + 5 + 4 + 4 = 19 이므로 사용된 글자는 모두 19개입니다.

1부터 1,000까지 영어로 썼을 때는 모두 몇 개의 글자를 사용해야 할까요?

참고: 빈 칸이나 하이픈('-')은 셈에서 제외하며, 단어 사이의 and 는 셈에 넣습니다.
   예를 들어 342를 영어로 쓰면 three hundred and forty-two 가 되어서 23 글자,
   115 = one hundred and fifteen 의 경우에는 20 글자가 됩니다.
'''
def number(N):
    if N == 1000:
        return 'onethousand'
    elif N >= 100:
        if N % 100 == 0:
            return number(N/100)+'hundred'
        else:            
            return number(N/100)+'hundredand' + number(N%100)
    elif N >= 20:    
        return ['twenty', 'thirty', 'fourty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninty'][N/10-2]+number(N%10)
    else:
        return ['', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'][N]


print sum([len(number(n)) for n in range(1, 1001)])

21124


In [75]:
 print ['twenty', 'thirty', 'fourty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninty'][3]

fifty


먼저 숫자를 영어로 출력할number()라는 함수를 만들어줍니다.
1000일경우에는 따로 onethousand를 출력하고 100이상일 경우에는 두가지 상황으로 나누어 100으로 나누어 떨어질 경우 100으로 나누어 일의 자리숫자가 나오면 영어와 뒤에 hundred가 붙어 반환되게 하엿고 100으로 나누어 떨어 지지 않는 경우 100으로 나눈 일의자리 숫자의 영어와 'hundredand'를 붙이고 뒤에 십의 자리는 number()의 인자로 넘겨줍니다.
그리고 20보다 큰 숫자가 들어올 경우는 10으로 나누어 몫을 앞의 십의자리 리스트를 참조하게 하였습니다. 그 후 10으로 나눈 나머지는 number()함수의 인자로 보냅니다.
일의 자리가 들어올 경우는 0부분을 비워놓고 19 즉 nineteen까지 타이핑 해두었습니다.
마지막으로 리스트 내포를 이용해 1부터 1000까지 for문을 이용해 number()함수에 넣고 len()을 사용해 숫자를 구해 리스트를 만들고 그 리스트를 sum()을 사용해 모두 더했습니다.

In [108]:
'''9. 다음과 같이 삼각형 모양으로 숫자를 배열했습니다.

3
7 4
2 4 6
8 5 9 3 

삼각형의 꼭대기부터 아래쪽으로 인접한 숫자를 찾아 내려가면서 합을 구하면, 위의 그림처럼 3 + 7 + 4 + 9 = 23 이 가장 큰 합을 갖는 경로가 됩니다.

다음 삼각형에서 합이 최대가 되는 경로를 찾아서 그 합을 구하세요.

75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23 

참고: 여기서는 경로가 16384개밖에 안되기 때문에, 모든 경로의 합을 일일이 계산해서 답을 구하는 것이 가능합니다.
하지만 67번 문제에는 100층짜리 삼각형 배열이 나옵니다. 그런 경우에는 좀 더 현명한 풀이 방법을 찾아야겠지요.
'''
s = '''75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23'''
# 맨아래층의 바로 위층 부터 자신의 아래층 중 좌/우 값 중 큰 값을 더하면서
# 맨 위층까지 올라가면, 맨 위층 값이 젤 큰 값

vals = []
for line in s.split('\n'):
    vals.append([int(num) for num in line.split()])

for row in range(13, -1, -1):
    for col in range(row+1):
        vals[row][col] += max(vals[row+1][col], vals[row+1][col+1])
print vals[0][0]


1074


문제를 거꾸로 생각하여 제일 아래에서 두 번째줄부터 시작 하여 아랫줄 두값중 큰값을 더해 위층 까지 올라가는 형식으로 풀었습니다.
먼저 문자열 s를 split('\n')을 사용해 줄별로 나누었습니다.
그후 빈 vals 리스트에 줄별로 나눈 문자열들을 정수형으로 바꾸어 리스트 형식으로 vals리스트에 추가해 줍니다. 그후 제일 아래 바로 윗 줄부터 아래에 좌우 값중 큰값을 자신에게 더하는 형식으로 제일 위인  vals[0][0]까지 값을 구하고 이를 출력합니다.

In [113]:
'''10. 다음은 달력에 관한 몇 가지 일반적인 정보입니다 (필요한 경우 좀 더 연구를 해 보셔도 좋습니다).
•1900년 1월 1일은 월요일이다.
•4월, 6월, 9월, 11월은 30일까지 있고, 1월, 3월, 5월, 7월, 8월, 10월, 12월은 31일까지 있다.
•2월은 28일이지만, 윤년에는 29일까지 있다.
•윤년은 연도를 4로 나누어 떨어지는 해를 말한다. 하지만 400으로 나누어 떨어지지 않는 매 100년째는 윤년이 아니며, 400으로 나누어 떨어지면 윤년이다

20세기 (1901년 1월 1일 ~ 2000년 12월 31일) 에서, 매월 1일이 일요일인 경우는 총 몇 번입니까?
'''

import calendar
sol=0
for i in range(1901,2001):
    for m in range(1,13):
        if calendar.weekday(i,m,1)==6:
            sol=sol+1
print sol



171


먼저 calendar 모듈을 불러옵니다. 그 이후에 이중 for문을 돌려 1901년부터 2000년그리고 1월부터 12월 까지 마지막은 1일로 두고 돌려줍니다. calendar 모듈에서 일요일 일 때는 6이라는 값을 반환해 줍니다. 그러므로 이중포문에 6이 있으면 카운트를 하나씩 올려 결국 171이 나옵니다 

In [119]:
'''11. n! 이라는 표기법은 n × (n − 1) × ... × 3 × 2 × 1을 뜻합니다.

예를 들자면 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800 이 되는데,
여기서 10!의 각 자리수를 더해 보면 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27 입니다.

100! 의 자리수를 모두 더하면 얼마입니까?
'''
i=1
for x in range(1,101):
    i=i*x

i=str(i)
n=0
for x in i:
    n=n+int(x)
print n

<type 'str'>
648


먼저 100팩토리얼을 for문을 통해 구하고 이를 문자열로 바꾸어 for문으로 숫자로된 문자를 하나씩 출력 합니다. 여기서 추출한 값은 문자이므로 int()를 사용해 정수형으로 바꾸어 더해주어 값을 출력합니다.

In [140]:
'''12. n의 약수들 중에서 자신을 제외한 것의 합을 d(n)으로 정의했을 때,
서로 다른 두 정수 a, b에 대하여 d(a) = b 이고 d(b) = a 이면 
a, b는 친화쌍이라 하고 a와 b를 각각 친화수(우애수)라고 합니다. 

예를 들어 220의 약수는 자신을 제외하면 1, 2, 4, 5, 10, 11, 20, 22, 44, 55, 110 이므로 그 합은 d(220) = 284 입니다.
또 284의 약수는 자신을 제외하면 1, 2, 4, 71, 142 이므로 d(284) = 220 입니다.
따라서 220과 284는 친화쌍이 됩니다. 

10000 이하의 친화수들을 모두 찾아서 그 합을 구하세요.
'''


def sum_y(number):
    
    result = 0
        
    for x in xrange(1,number+1):
        if number%x == 0:
            result += x
    return result
            
    


def findnum(number):
        x = sum_y(number) - number
        if sum_y(x) - x == number:
            return x
        return None

s=[]

for i in xrange(1, 10001):
    x = findnum(i)
    if (x != None) and (x != i) :
        s.append(x)
       
        
print sum(s)


31626


먼저 sum_y()함수는 들어온 인자값의 약수들을 모두 더해주어 반환해주는 함수입니다.
findnum()함수에서 인자값의 약수를 sum_y()을 사용해 모두 더해준뒤 자기자신의 값을 빼서 x에 저장해 줍니다. 만약에 그 x값을 똑같이 sum_y()에 인자값으로 넣고 자신 x값을 뺀 값이 처음 넣은 값과 같다면 이는 친화수 이므로 x값을 반환하는 함수입니다.

for문을 사용해 1부터 10000까지의 친화수를 검사한 후 친화수 이면 리스트 s에 추가하여주고 마지막에 sum()을 사용해 모두 더해줍니다.

In [145]:
'''13번 문제 이전 Assignment 3 (Assignment 4가 아님)의 마지막 문제는 웹 URL로 지정된 웹페이지를 문자열로 가져와
모든 HTML 태그 및 CSS와 Javascript를 제외한 순수 텍스트를 얻어내고 그 안에 존재하는 단어를 추출하여 각 단어들에 대해 출현빈도를 
사전형태({'world': 2, 'hello': 1, 'python': 1})로 저장하여 출력하는 것이었다.
이번에는 Assignment 3을 다시 확장/변형하여 다음과 같은 조건을 만족하도록 구현하시오.'''
# -*- coding: utf-8 -*-
from operator import itemgetter
import string
import urllib
class WebWordsFrequency:
    def __init__(self,*arg):
        self.Arg=list(arg)
        
    def addUrl(self,other):
        w=self.Arg
        w.append(other)
        
    def removeUrl(self,other):
        w=self.Arg
        w.remove(other)
        
    def listUrls(self):
        print '\n'.join(self.Arg)
        
    def getWordsFrequency(self):
        a={}
        if len(self.Arg) == 0:
            print a
            return a
            
        w1 = self.Arg
        for x in range(len(self.Arg)):
            s = urllib.urlopen(w1[x]).read()
            start_num1='<'
            finish_num2='>'
            f=[]
            a = False
            range_k = range(len(s))
            for k in range_k:
                if s[k] == start_num1:
                    a = True
                elif s[k]==finish_num2:
                    a = False
                elif a == True:
                    continue
                else:
                    f.append(s[k])
            b = ''.join(f)
        
            g=[]
            c=False
            start_num3='{'
            finish_num4='}'
            range_l = range(len(b))
            for l in range_l:
                if b[l] == start_num3:
                    c = True
                elif b[l]==finish_num4:
                    c = False
                elif c == True:
                    continue
                else:
                    g.append(b[l])
            d = ''.join(g)
        
            h=[]
            e=False
            start_num5='.'
            finish_num6=';'
            range_x = range(len(d)) 
            for l in range_x:
                if d[l] == start_num5:
                    e = True
                elif d[l]==finish_num6:
                    e = False
                elif e == True:
                    continue
                else:
                    h.append(d[l])
            w = ''.join(h)
            t=string.punctuation
            new_w=w.translate(None,t)
        
            D={}
            L=new_w.split()
            for word in L:
                if word in D:
                    D[word]+=1
                else:
                    D[word]=1
        
            for key in D.keys():
                print "%s : %s" % (key, D[key]),
               
        #return D
 
    def getMaxFreqencyWords(self):
        if len(self.Arg) == 0:
            print None
            return None
        else:
            w1 = self.Arg
            for x in range(0,len(self.Arg)):
                s = urllib.urlopen(w1[x]).read()
                start_num1='<'
                finish_num2='>'
                f=[]
                a = False
                range_k = range(len(s))
                for k in range_k:
                    if s[k] == start_num1:
                        a = True
                    elif s[k]==finish_num2:
                        a = False
                    elif a == True:
                        continue
                    else:
                        f.append(s[k])
                b = ''.join(f)
            
                g=[]
                c=False
                start_num3='{'
                finish_num4='}'
                range_l = range(len(b))
                for l in range_l:
                    if b[l] == start_num3:
                        c = True
                    elif b[l]==finish_num4:
                        c = False
                    elif c == True:
                        continue
                    else:
                        g.append(b[l])
                d = ''.join(g)
            
                h=[]
                e=False
                start_num5='.'
                finish_num6=';'
                range_x = range(len(d)) 
                for l in range_x:
                    if d[l] == start_num5:
                        e = True
                    elif d[l]==finish_num6:
                        e = False
                    elif e == True:
                        continue
                    else:
                        h.append(d[l])
                w = ''.join(h)
                t=string.punctuation
                new_w=w.translate(None,t)
            
                D={}
                L=new_w.split()
                for word in L:
                    if word in D:
                        D[word]+=1
                    else:
                        D[word]=1
                t1=(sorted(D.iteritems(),key=itemgetter(1),reverse=True))
                print t1[0][0]
                for x in range(1, len(t1)):
                    if t1[0][1] == t1[x][1] :
                        print t1[x][0]
                #return t1
                continue

w1=WebWordsFrequency('http://www.cnn.com', 'http://www.times.com', 'https://www.amazon.com')
w2 = WebWordsFrequency('http://www.cnn.com', 'http://www.times.com') 
w3 = WebWordsFrequency() 
w1.addUrl('https://github.com')
w3.addUrl('http://stackoverflow.com')
w1.removeUrl('http://www.cnn.com')
w2.removeUrl('http://www.times.com')
w1.listUrls()
print '\n\n사전\n'
w1.getWordsFrequency()
print '\n\n최다출현\n'
w2.getMaxFreqencyWords()


http://www.times.com
https://www.amazon.com
https://github.com


사전

Sell : 1 Van : 1 founder : 1 windowet0forvar : 1 Take : 1 ERIC : 1 ArtsBeat : 1 Honda : 1 Ridley : 1 to : 22 Now : 4 Condemned : 1 Retreat : 1 Heaven’ : 1 ‘Transparent’ : 1 Getaway : 1 Old : 1 Sessanta : 1 little : 1 Concert : 1 Chemical : 1 Draws : 1 Rocket : 1 Campaign : 2 Less : 1 SHANE : 1 Related : 1 Defense : 1 Resort : 1 Debate : 2 Sensible : 1 force : 2 be : 1 Merger : 1 asetPageViewNameaddPageActionsetCustomAttributefinishedaddToTraceinlineHitnoticeErroroafunctionennnnnvar : 1 Should : 1 270 : 1 Matthew : 1 State : 2 Strange : 1 Slowly : 1 Finding : 1 Drawn : 1 Brace : 1 Nytimes : 1 Responses : 1 Subscriptions : 1 Memoir : 1 Assault : 1 rvar : 1 uvar : 1 America : 1 True : 1 Ethicist : 1 Clan : 1 Bill : 1 Age : 2 Changed : 1 hundreds : 1 Education : 2 Green : 1 Our : 2 David : 1 My : 1 borrowers : 1 Birds : 1 Business : 1 Deal : 1 Absolutely : 2 Intelligence : 1 Back : 1 Dead : 1 Your : 1 Born : 1 JComments :

class내부에 인자를 받을init 함수를 정의를 하였는 데 이때 인자를 받는 것에 대한 수는 정해지지 않았으므로  몇 개가 들어올지 몰라서 가변인수를 받게 하였습니다.  addUrl의 경우 init에 있는 값에 인자를 추가를 할 수 있게 정의 하였고 removeUrl을 이용하여 그 값들 중 지정된 값을 삭제하고 나머지 만 남게 하였습니다.  listUrl을 이용하여 추가하고 삭제한것에 대한 최종 값을 저장을 하였고 getwordsfrequency를 이용하여 이전의 과제들에서 순순한 텍스트만 받아와서 출력하는 것을 함수로서 정의를 하였으며 getMaxfrequency의 경우 가장 많은 빈도수를 보여주는 단어를 출력을 하되 빈도수가 같은것은 전부 출력이 되게 만들었습니다.

In [143]:
'''14번 문제 13번 문제에서 정의한 WebWordsFrequency 클래스를 상속하여
OrderedWebWordsFrequency 클래스를 정의하고 슈퍼클래스에 정의된
getWordsFrequency() 메소드를 오버라이드 하여 단어 출현 빈도를 내림 차순으로 정렬하여 리스트로 출력하시오'''
# -*- coding: utf-8 -*-
from operator import itemgetter
import string
import urllib
import urllib2
class WebWordsFrequency:
    def __init__(self,*arg):
        self.Arg=list(arg)
        self.directory = {}
        self.l = []
        for url in arg:
            if url not in self.l:
                self.l.append(url)
        
    def addUrl(self,other):
        w=self.Arg
        w.append(other)
        
    def removeUrl(self,other):
        w=self.Arg
        w.remove(other)
        
    def listUrl(self):
        print self.Arg
        
    def getWordsFrequency(self):
        a={}
        if len(self.Arg) == 0:
            print a
            return a
            
        w1 = self.Arg
        for x in range(len(self.Arg)):
            s = urllib.urlopen(w1[x]).read()
            start_num1='<'
            finish_num2='>'
            f=[]
            a = False
            range_k = range(len(s))
            for k in range_k:
                if s[k] == start_num1:
                    a = True
                elif s[k]==finish_num2:
                    a = False
                elif a == True:
                    continue
                else:
                    f.append(s[k])
            b = ''.join(f)
        
            g=[]
            c=False
            start_num3='{'
            finish_num4='}'
            range_l = range(len(b))
            for l in range_l:
                if b[l] == start_num3:
                    c = True
                elif b[l]==finish_num4:
                    c = False
                elif c == True:
                    continue
                else:
                    g.append(b[l])
            d = ''.join(g)
        
            h=[]
            e=False
            start_num5='.'
            finish_num6=';'
            range_x = range(len(d)) 
            for l in range_x:
                if d[l] == start_num5:
                    e = True
                elif d[l]==finish_num6:
                    e = False
                elif e == True:
                    continue
                else:
                    h.append(d[l])
            w = ''.join(h)
            t=string.punctuation
            new_w=w.translate(None,t)
        
            D={}
            L=new_w.split()
            for word in L:
                if word in D:
                    D[word]+=1
                else:
                    D[word]=1      
        return D
    def parserURL(directory, source):
        soup = BeautifulSoup(source, "html.parser")
        pureWords = soup.text.replace("//","")

        pureWords = WebWordsFrequency.deleteCDATA(pureWords)

        scrapList = string.punctuation
        for i in range(len(scrapList)):
            pureWords = pureWords.replace(scrapList[i],'')

        wordCount = pureWords.split()

        
        for word in wordCount:
            if word in directory:
                directory[str(word.encode('utf-8'))] +=1
            else:
                directory[str(word.encode('utf-8'))] = 1

        stopFilter = ['a','and','are','as','at','be','but','by','for', 
                  'if','in','into','is','it','no','not','of','on',
                  'or','s','such','t','that','the','their','then',
                  'there','these','they','this','to','was','will',
                  'with']

        for word in directory.keys():
            if word in stopFilter:
                del directory[word]
                
    def getMaxFreqencyWords(self):
        if len(self.Arg) == 0:
            print None
            return None
        else:
            w1 = self.Arg
            for x in range(0,len(self.Arg)):
                s = urllib.urlopen(w1[x]).read()
                start_num1='<'
                finish_num2='>'
                f=[]
                a = False
                range_k = range(len(s))
                for k in range_k:
                    if s[k] == start_num1:
                        a = True
                    elif s[k]==finish_num2:
                        a = False
                    elif a == True:
                        continue
                    else:
                        f.append(s[k])
                b = ''.join(f)
            
                g=[]
                c=False
                start_num3='{'
                finish_num4='}'
                range_l = range(len(b))
                for l in range_l:
                    if b[l] == start_num3:
                        c = True
                    elif b[l]==finish_num4:
                        c = False
                    elif c == True:
                        continue
                    else:
                        g.append(b[l])
                d = ''.join(g)
            
                h=[]
                e=False
                start_num5='.'
                finish_num6=';'
                range_x = range(len(d)) 
                for l in range_x:
                    if d[l] == start_num5:
                        e = True
                    elif d[l]==finish_num6:
                        e = False
                    elif e == True:
                        continue
                    else:
                        h.append(d[l])
                w = ''.join(h)
                t=string.punctuation
                new_w=w.translate(None,t)
            
                D={}
                L=new_w.split()
                for word in L:
                    if word in D:
                        D[word]+=1
                    else:
                        D[word]=1
                t1=(sorted(D.iteritems(),key=itemgetter(1),reverse=True))
                for x in range(1, len(t1)):
                    if t1[0][1] == t1[x][1] :
                        print t1[0][0],t1[x][0]
                        break
                    elif t1[x][1] > t1[x][1] :
                        print t1[x][0]
                        break
                    else:
                        print 'error'
                        break
                return t1
                continue

class OrderedWebWordsFrequency(WebWordsFrequency):
    def __init__(self,*arg):
        WebWordsFrequency.__init__(self,*arg)
    def getWordsFrequency(self,reverse=False):
        t1=(sorted(WebWordsFrequency.getWordsFrequency(self).iteritems(),key=itemgetter(1),reverse=not(reverse)))
        for key in range(0,len(t1)):
                print "%s :%s " % t1[key],
    

w4 = OrderedWebWordsFrequency('http://www.times.com' ,'https://www.amazon.com',)

w4.getWordsFrequency()
print '\n\n'
w4.getWordsFrequency(reverse=True)


window :59  P :22  var :12  if :11  more :8  See :7  Nav :6  Services :5  Amazon :4  on :4  Home :4  params :4  in :4  Movies :4  d :3  o :3  Most :3  function :3  with :3  typeof :3  Your :3  Digital :3  Games :3  TV :3  opts :3  isSafari :2  to :2  Sellers :2  0 :2  auiPreloadIndex :2  SearchJS :2  Prime :2  return :2  Cards :2  amp :2  For :2  Wished :2  for :2  new :2  object :2  END :2  of :2  Best :2  Get :2  else :2  an :2  Video :2  ue :2  Payment :2  AmazonSell :2  experienceStr :2  Ship :1  aPageStart :1  Web :1  Gourmet :1  Handmade :1  Improvement :1  catch :1  Happiness :1  Supplies :1  loading :1  wafunction :1  issCorpus :1  fabccatchdreturn :1  menu :1  gaifd :1  alwaysRaisedyes :1  Handpicked :1  UsCareersAbout :1  cfevar :1  invocation :1  Wine :1  suggestionText :1  Travel :1  AmazonBecome :1  Organizers :1  Grocery :1  UsBecome :1  cfunction :1  BeautyBar :1  KflAmazonUIPageJSnew :1  MP3 :1  FJnH1glbeforeunloadfunction1E4var :1  Products :1  now :1  Audible :1  Shop

oderedwebfrequency()에 webwordfrequency()를 상속을 받고 webwordfrequency()에서 정의 되어있는 함수 getwordfrequency()를 정보를 받아 그것의 키값을 비교하여 sorted()함수로 내림차순 정렬 하였습니다. 또 reverse 가 거짓일 때 내림차순으로 출력해야 하므로 sorted의 3번째 칸에 reverse값에 not을 붙여 true로 만들어 주었고 reverse를 true로 받을 경우 false로 인식하게 하였습니다.

In [144]:
'''15번 문제 

다음과 같은 코딩이 가능하도록 OrderedWebWordsFrequency 안에 반복자와 관련된 메소드를 추가하시오.


for i in w4:
 print i
 ('site, 12)
 ('science', 11)
 ('hello', 8)
 ('world', 2)
 ('program', 1)
 ('python', 1)
'''
class OrderedWebWordsFrequency(WebWordsFrequency):
    def __init__(self,*arg):
        WebWordsFrequency.__init__(self,*arg)
    def getWordsFrequency(self,reverse=False):
        self.newList = []
        t1=(sorted(WebWordsFrequency.getWordsFrequency(self).iteritems(),key=itemgetter(1),reverse=not(reverse)))
        for key in range(0,len(t1)):
            self.newList.append(t1[key])
            
               
        
        

    def __iter__(self):
        self.getWordsFrequency()
        self.count = -1
        return self


    def next(self):
        self.count += 1
        if self.count < len(self.newList):
            return self.newList[self.count]
        else:
            raise StopIteration



w4 = OrderedWebWordsFrequency('http://www.times.com','https://www.amazon.com' )
for i in w4:
    print i

('window', 59)
('P', 25)
('if', 14)
('var', 12)
('more', 8)
('See', 7)
('typeof', 6)
('Nav', 6)
('object', 5)
('Services', 5)
('ue', 5)
('Amazon', 4)
('on', 4)
('Home', 4)
('params', 4)
('with', 4)
('d', 3)
('of', 3)
('o', 3)
('to', 3)
('function', 3)
('Your', 3)
('Digital', 3)
('in', 3)
('Movies', 3)
('opts', 3)
('isSafari', 2)
('0', 2)
('auiPreloadIndex', 2)
('SearchJS', 2)
('Prime', 2)
('return', 2)
('Cards', 2)
('for', 2)
('new', 2)
('END', 2)
('Gift', 2)
('else', 2)
('an', 2)
('and', 2)
('Get', 2)
('Payment', 2)
('Games', 2)
('AmazonSell', 2)
('experienceStr', 2)
('TV', 2)
('Ship', 1)
('aPageStart', 1)
('Unlocked', 1)
('Gourmet', 1)
('Fitness', 1)
('Handmade', 1)
('Improvement', 1)
('catch', 1)
('Sonicare', 1)
('Happiness', 1)
('Supplies', 1)
('loading', 1)
('wafunction', 1)
('issCorpus', 1)
('fabccatchdreturn', 1)
('menu', 1)
('gaifd', 1)
('alwaysRaisedyes', 1)
('Handpicked', 1)
('UsCareersAbout', 1)
('Sellers', 1)
('cfevar', 1)
('invocation', 1)
('Wine', 1)
('suggestionText', 1)

반복자를 사용하기위해서 __iter__와 next()에 대해 재정의 하였고 
next()에서 0부터 호출하기 위해 iter()에서 counter 값을 -1로 주었습니다.
next는 멤버 변수 count를 1증가 시키고 count가 getWordsFrequency()로 만든 newList의 개수 보다 작을 경우 계속 저장된값을 출력 하나씩 출력하도록 하였습니다.

#고찰
마지막과제까지 끝내고 보니 시작 할 때와 다르게 파이썬에 대해 제법 많은 것을 알게 되었습니다. 항상 과제를 할 때 느끼는 것이지만 역시 복습 하지 않으면 배웠던 것을 항상 잊게 되는 것 같습니다. 오일러 문제도 어려운게 많았고 교수님이 내주신 문제도 간혹가다 어려워서 이곳 저곳 많이 알아보고 도움도 받았습니다. 처음에는 이것을 자바로 코딩하면 금방 할수 있을 것 같았는데 이제는 파이썬이 더 손에 익어 파이썬으로 프로그래밍을 하고 싶어졌습니다. 그동안 수고하셨습니다!